lib/dockerregistry/inventory.go

```diff
@@ -1353,7 +1349,8 @@ func (sc *SyncContext) ReadRegistries(
 			}
 			// When we're done processing this node (req), decrement the
 			// semaphore.
-			wg.Add(-1)
+			reqRes.Errors = Errors{}
```

Comment:
Do we absolutely have to move this block from line 1321?

Reply:
Yes, we do. This is because in that recurse code block (line 1321 in this pr) we are creating new requests and pushing them into the requests channel. However, you can imagine the scenario where we push a result into the requestsResults channel which results in the WaitGroup being drained to 0. As soon as the semaphore is drained, both the requests channel and the requestResults channel are closed. This causes a panic error because the recurse block is trying to push requests into a channel that has just been closed.

Reply:
Ah, that's right. I think the much simpler explanation is, because we moved the semaphore decrement to ExecRequests, we need to move the push into the requestResults channel to where the old semaphore decrement in *this* function *used to be*, which is the location here at line 1352 (clearly shown in the diff). Otherwise if we leave it as-is, then it's as if we are doing `wg.Add(1)` back up at line 1321, which is clearly wrong.

Thanks!